src/internal/eval-tidy.c

```diff
@@ -340,24 +340,76 @@ sexp* rlang_new_data_mask_compat(sexp* bottom, sexp* top, sexp* parent) {
 sexp* rlang_as_data_mask_compat(sexp* data, sexp* parent) {
   return rlang_as_data_mask(data);
 }
-
+// Similar to Rf_findFun(r_sym("~"), env), but skips the mask environment.
+// In theory we need a version that does Rf_findVar(r_sym("~"), env) for
+// the cases where `~` is referenced directly (instead of called as a fun),
+// but there is no point to it since we have no mechanism to trigger it like
+// we do when the rlang mask tilde evaluates.
+//
+// Adapted from findFun (R/trunk/src/main/envir.c)
+
+static sexp* findTilde(sexp* env) {
```

Comment:
This function seems complicated. Can you use `findFun(tilde_sym, env)` with `env` the parent of the mask?

I think we can just check that env is an rlang mask, and if so call findFun with the parent. This allows the user data masks to override `~`, which I guess makes sense.

Reply:
Agree this function isn't ideal.  I haven't had a chance to think about this one but will circle back when I do.

Reply:
I can also finish this up if you would like. Otherwise I think we'll release in one or two weeks so there's still a bit of time.

Reply:
In untested pseudo-code, are you thinking of something like:

```
env               // this is `caller_frame` from rlang_tilde_eval
while((tilde = findFun('~', env)) == tilde_fn) {
  env = r_env_parent(r_env_find_anywhere(env, quo_mask_flag_sym));
}
tilde
```

i.e. if the found tilde is the rlang mask one, find the mask env with quo_mask_flag_sym, and repeat until a tilde is found that isn't the rlang mask one?

Reply:
I was thinking of just calling findFun in the immediate parent of the mask and findFun would do the looping. Or am I missing something?

Reply:
It could be me.  I assumed we needed the looping because of things like:

```
`~` <- function(...) stop('success!')
eval_tidy(quo(eval_tidy(~1)))
```

Where the mask frames set up by nested `eval_tidy` calls would interfere sequentially.  I'll admit though my grasp of the rlang internals is not complete.

Reply:
It's a good case to think about. From a cursory thinking, the tidyeval-tilde will call another tidyeval-tilde, which should then repeat the process and find the lexical tilde?

In any case, I don't want this code to be too complicated just to support very unlikely edge cases.

Reply:
I think this causes an infinite loop because we can't just eval the tilde expression in the enclosure of the mask environment as the tilde expression may be referencing elements in the calling frame.  If we eval in the calling frame, then it's an infinite loop.

```
`~` <- function(x, y) x + y
eval_tidy(quo({
  a <- 1
  a ~ 41
}))
```

This variation causes an infinite loop:

```
static sexp* base_tilde_eval(sexp* tilde, sexp* quo_env) {
  if (r_f_has_env(tilde)) {
    return tilde;
  }
  // Inline the `~` because data masks override it to make
  // quosures self-evaluate

  sexp* tilde_sym = KEEP(r_sym("~"));
  sexp* tilde_found_fn = KEEP(Rf_findFun(tilde_sym, quo_env));

  tilde = KEEP(r_new_call(tilde_found_fn, r_node_cdr(tilde)));
  tilde = KEEP(r_eval(tilde, quo_env));

  // Undo inlining but only if the result of evaluating it is still an
  // unevaluated call with the same "tilde"

  if (TYPEOF(tilde) == LANGSXP && CAR(tilde) == tilde_found_fn) {
    r_node_poke_car(tilde, r_tilde_sym);
  }
  FREE(4);
  return tilde;
}
```

This variation fails to find `a`:

```
static sexp* base_tilde_eval(sexp* tilde, sexp* quo_env) {
  if (r_f_has_env(tilde)) {
    return tilde;
  }
  // Inline the `~` because data masks override it to make
  // quosures self-evaluate

  sexp* tilde_sym = KEEP(r_sym("~"));
  sexp* tilde_found_fn = KEEP(Rf_findFun(tilde_sym, quo_env));
  sexp* mask_env = r_env_find_anywhere(quo_env, quo_mask_flag_sym);
  if (mask_info(mask_env).type == RLANG_MASK_QUOSURE) {
    quo_env = r_env_parent(mask_env);
  }
  tilde = KEEP(r_new_call(tilde_found_fn, r_node_cdr(tilde)));
  tilde = KEEP(r_eval(tilde, quo_env));

  // Undo inlining but only if the result of evaluating it is still an
  // unevaluated call with the same "tilde"

  if (TYPEOF(tilde) == LANGSXP && CAR(tilde) == tilde_found_fn) {
    r_node_poke_car(tilde, r_tilde_sym);
  }
  FREE(4);
  return tilde;
}
```

Perhaps you meant something else I'm not grasping?

Either way, I agree this is not worth a huge amount of effort or code, although it should probably be documented.  I did this mostly b/c I'm interested in understanding the internals better.